File: algorithms/1_easy/M/00169_majority_element/boyer_moore_voting.cpp
class Solution {
public:
    static int majorityElement(const vector<int>& nums) {
        int majority_candidate = 0;
        int count = 0;

        for (const auto num : nums) {
            if (count == 0) {
                majority_candidate = num;
                count = 1;
            } else if (majority_candidate == num) {
                count += 1;
            } else {
                count -= 1;
            }
        }

        return majority_candidate;
    }
};

File: algorithms/1_easy/M/00169_majority_element/boyer_moore_voting.rs
impl Solution {
    pub fn majority_element(nums: Vec<i32>) -> i32 {
        let mut majority_candidate = 0_i32;
        let mut count = 0_i32;

        for num in nums.into_iter() {
            if count == 0 {
                majority_candidate = num;
                count = 1;
            } else if majority_candidate == num {
                count += 1;
            } else {
                count -= 1;
            }
        }

        majority_candidate
    }
}
